lib/mongoid/document.rb

```diff
@@ -353,7 +353,7 @@ def instantiate(attrs = nil, selected_fields = nil)
       #
       # @since 1.0.0
       def _types
-        @_type ||= (descendants + [ self ]).uniq.map(&:to_s)
+        @_type = (descendants + [ self ]).uniq.map { |type| type.discriminator_value || type.to_s }
```

Comment:
What happened to `||=` here?

Reply:
And shouldn't discriminator_value always be defined?

Reply:
The `||=` was removed on purpose... Emily and I worked through this together... There was a problem with the test beginning on line 132 of `discriminator_value_spec.rb`... Basically when `count` is called a `View` is created which contains a specific filter i.e. `{ "_type" => "Circle` }`. The problem is when changing the discriminator key/value and calling count again, this filter is not changed because the `@type` was cached, so the wrong count would be returned. Removing the `||=` allows the filter to be calculated when it needs to be.

Reply:
The `discriminator_value` is not always defined, I've recently discovered. Specifically in the test in `document_spec.rb` line 108... In this test a descendant of `Person` is created using `Class.new(Person)`, and in this new class the discriminator value is null and the test expects something of the form "<Class:.....>".

Reply:
The `@_type` is a cache, so if you need the value to be recalculated every call you should remove the assignment to `@_type` and the clear in traversable.rb.

Reply:
How do I clear the `@_type` cache?

Reply:
I added a _clear_types method that clears the `@_type` cache.